appserver/web/web-glue/src/main/java/com/sun/enterprise/web/WebContainer.java

```diff
@@ -618,6 +620,27 @@ public void event(Event<?> event) {
             loadDefaultWebModulesAfterAllAppsProcessed();
         } else if (event.is(PREPARE_SHUTDOWN)) {
             isShutdown = true;
+        } else if (event.is(Deployment.DEPLOYMENT_SUCCESS)) {
+            ApplicationInfo applicationInfo = (ApplicationInfo) event.hook();
+            if (applicationInfo.getSource().getArchiveMetaData("commandparams", DeployCommandParameters.class).isRedeploy()) {
+                for (VirtualServer vs : getVirtualServers()) {
+                    if (ADMIN_VS.equals(vs.getName())) {
+                        continue;
+                    }
+
+                    if (vs.getDefaultWebModuleID().equals(applicationInfo.getName())) {
+                        try {
+                            // since this event is asynchronous, wait application module fully loaded,
+                            Thread.sleep(500);
```

Comment:
Not keen on this: it kind of defeats the purpose of using the event listener - if you're reacting to an event by waiting, why not react to a different event?

If the application isn't fully loaded at this point and we need it to be, I'd suggest we need to use a different event.
There's an `APPLICATION_LOADED` or `APPLICATION_STARTED` event we could use which seems more in line with what you're trying to do here.

Reply:
As our discussion, I have created a new event to notify the deployment command is finished and listen to it to resolve the missing application module while updating the host.